websocket_browser.go

```diff
@@ -5,26 +5,73 @@ package websocket
 import (
 	"context"
 	"errors"
+	"fmt"
 	"syscall/js"
 
 	"github.com/libp2p/go-libp2p-core/transport"
+	tptu "github.com/libp2p/go-libp2p-transport-upgrader"
 	ma "github.com/multiformats/go-multiaddr"
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
-func (t *WebsocketTransport) maDial(ctx context.Context, raddr ma.Multiaddr) (manet.Conn, error) {
+// WebsocketTransport is the actual go-libp2p transport
+type WebsocketTransport struct {
+	Upgrader *tptu.Upgrader
+	Config   *WebsocketConfig
+}
+
+func New(u *tptu.Upgrader) *WebsocketTransport {
+	return &WebsocketTransport{
+		Upgrader: u,
+		Config:   DefaultWebsocketConfig(),
+	}
+}
+
+// NewWithOptions returns a WebsocketTransport constructor function compatible
+// with the libp2p.New host constructor. Currently in the browser no options
+// are supported.
+func NewWithOptions(opts ...Option) func(u *tptu.Upgrader) *WebsocketTransport {
+	c := DefaultWebsocketConfig()
+
+	// Apply functional options.
+	for _, o := range opts {
+		o(c)
+	}
+
+	return func(u *tptu.Upgrader) *WebsocketTransport {
+		t := &WebsocketTransport{
+			Upgrader: u,
+			Config:   c,
+		}
+		return t
+	}
+}
+
+func (t *WebsocketTransport) maDial(ctx context.Context, raddr ma.Multiaddr) (mnc manet.Conn, err error) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
What is this used for?

Reply:
IIRC, we were concerned about some panics in the underlying library. I don't know if those have been fixed.